In [2]:
#Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [43]:
pd.options.display.max_columns = None

In [46]:
pd.options.display.max_rows = 10

In [3]:
#Importing orders_products_merged_updated2.pkl
newpath = 'preparedData'

In [5]:
ords_prods_merged_df2= pd.read_pickle(os.path.join(newpath,'orders_products_merged_updated2.pkl'))

In [6]:
#Importing customers.csv 
path = 'OriginalData'

In [7]:
customers_df= pd.read_csv(os.path.join(path,'customers.csv'))  

In [8]:
customers_df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
customers_df.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [10]:
customers_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [11]:
customers_df.shape

(206209, 10)

#4.Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

#Data Wrangling 

- Change 'surnam' to 'last_name'
- Change 'First Name' to 'first_name'
- Change 'STATE' to 'state'
- Change 'gender' to lowercase 'age'
- Change 'Age' to lowercase 'age'
- Change 'fam_status' to 'martial_status'
- Change 'n_dependants' to 'num_of_dependants'
- Check for missing values
- Check for mixed types
- Check for duplicates

In [12]:
# changing the name of columns 
customers_df.rename(columns={"First Name":"first_name", "Surnam":"last_name", "Gender":"gender", "STATE":"state",
                            "Age":"age","fam_status" :"martital_status","n_dependants":"num_of_dependants"}, inplace=True)

In [13]:
#Checking the renamed columns
customers_df.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependants martital_status  income  
0                  3         married  165665  
1                  0          single   59285  
2                  2         married   99568  
3                  0          single   42049  
4                  1         married   40374

# Checking for missing values



In [14]:
customers_df['first_name'].value_counts(dropna = False)

first_name
NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: count, Length: 208, dtype: int64

In [15]:
customers_df['last_name'].value_counts(dropna = False)

last_name
Hamilton      252
Randall       248
Lamb          243
Pennington    243
Barnett       242
             ... 
Poole         172
Bauer         166
Pearson       164
Payne         163
Jordan        162
Name: count, Length: 1000, dtype: int64

In [16]:
customers_df['gender'].value_counts(dropna = False)

gender
Male      104067
Female    102142
Name: count, dtype: int64

In [17]:
customers_df['state'].value_counts(dropna = False)

state
Florida                 4044
Colorado                4044
Illinois                4044
Alabama                 4044
District of Columbia    4044
Hawaii                  4044
Arizona                 4044
Connecticut             4044
California              4044
Indiana                 4044
Arkansas                4044
Alaska                  4044
Delaware                4044
Iowa                    4044
Idaho                   4044
Georgia                 4044
Wyoming                 4043
Mississippi             4043
Oklahoma                4043
Utah                    4043
New Hampshire           4043
Kentucky                4043
Maryland                4043
Rhode Island            4043
Massachusetts           4043
Michigan                4043
New Jersey              4043
Kansas                  4043
South Dakota            4043
Minnesota               4043
Tennessee               4043
New York                4043
Washington              4043
Louisiana               4043
Montana 

In [18]:
customers_df['age'].value_counts(dropna = False)

age
19    3329
55    3317
51    3317
56    3306
32    3305
      ... 
65    3145
25    3127
66    3114
50    3102
36    3101
Name: count, Length: 64, dtype: int64

In [19]:
customers_df['date_joined'].value_counts(dropna = False)

date_joined
9/17/2018     213
2/10/2018     212
4/1/2019      211
9/21/2019     211
12/19/2017    210
             ... 
9/1/2018      141
1/22/2018     140
11/24/2017    139
7/18/2019     138
8/6/2018      128
Name: count, Length: 1187, dtype: int64

In [20]:
customers_df['num_of_dependants'].value_counts(dropna = False)

num_of_dependants
0    51602
3    51594
1    51531
2    51482
Name: count, dtype: int64

In [21]:
customers_df['martital_status'].value_counts(dropna = False)

martital_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [ ]:
customers_df['income'].value_counts(dropna = False)

Observation on missing values

- first_name has 11,259 null has values. 
- No missing values in income.

# Checking for mixed types

In [ ]:

# looking in customer df to see if there are any mixed type
for col in customers_df.columns.tolist():
    weird = (customers_df[[col]].applymap(type) != customers_df[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (customers_df[weird]) > 0:
        print (col)

##Firstname column contains mixed type

In [22]:
customers_df.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
state                    0
age                      0
date_joined              0
num_of_dependants        0
martital_status          0
income                   0
dtype: int64

In [23]:
customers_df.loc[customers_df['first_name'].isna()]

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        num_of_dependants martital_status  income  
53                      2         married   41709  
73                      0          single   82518  
82                      3         married  117099  
99                      1         married  155673  
105                     3         married  151819  
...                   ...             ...     ...  
206038                  3         married   87783  
206044                  2         married  108488  
206090                  1         married   45275  
206162                  0          single   56325  
206171                  2         married   59222  

[11259 rows x 10 columns]

In [24]:
##Imputation of missing values 
##filling the missing values in the first_name column with 'unknown'
customers_df['first_name'].fillna('Unknown', inplace = True)

In [25]:
#rechecking the imputation of missing values
customers_df.isnull().sum()

user_id              0
first_name           0
last_name            0
gender               0
state                0
age                  0
date_joined          0
num_of_dependants    0
martital_status      0
income               0
dtype: int64

# Checking for Duplicates

In [26]:
customers_dups_df = customers_df[customers_df.duplicated()]


In [27]:
customers_dups_df

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_of_dependants, martital_status, income]
Index: []

Observation - No Duplicates were found.

#6Combine your customer data with the rest of your prepared Instacart data.

In [28]:
ords_prods_merged_df2.head(80)

Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0              0   2539329        1    prior             1   
1              1   2398795        1    prior             2   
2              2    473747        1    prior             3   
3              3   2254736        1    prior             4   
4              4    431534        1    prior             5   
..           ...       ...      ...      ...           ...   
75          1744    520620      120    prior             1   
76          1746   3273029      120    prior             3   
77          2781   1861943      185    prior             2   
78          2891   2040988      195    prior             1   
79          2898   1680569      195    prior             8   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
..                 ...                ...                     ...         ...   
75                   3                 11                     NaN         196   
76                   2                  8                    19.0         196   
77                   1                 12                    25.0         196   
78                   1                 14                     NaN         196   
79                   1                  9                     4.0         196   

    add_to_cart_order  ...  prices     busiest_day     busiest_days  \
0                   1  ...     9.0  Regularly busy   Regularly busy   
1                   1  ...     9.0  Regularly busy  Least busy days   
2                   1  ...     9.0  Regularly busy  Least busy days   
3                   1  ...     9.0      Least busy  Least busy days   
4                   1  ...     9.0      Least busy  Least busy days   
..                ...  ...     ...             ...              ...   
75                  2  ...     9.0  Regularly busy  Least busy days   
76                  2  ...     9.0  Regularly busy   Regularly busy   
77                 10  ...     9.0  Regularly busy     Busiest days   
78                  2  ...     9.0  Regularly busy     Busiest days   
79                  2  ...     9.0  Regularly busy     Busiest days   

   busiest_period_of_day  max_order    loyalty_flag  average_price  \
0         Average orders         10    New customer       6.367797   
1         Average orders         10    New customer       6.367797   
2            Most orders         10    New customer       6.367797   
3         Average orders         10    New customer       6.367797   
4            Most orders         10    New customer       6.367797   
..                   ...        ...             ...            ...   
75           Most orders          3    New customer       9.385714   
76        Average orders          3    New customer       9.385714   
77           Most orders          6    New customer       7.715000   
78           Most orders         58  Loyal customer       7.149730   
79           Most orders         58  Loyal customer       7.149730   

   spending_flag median_order_frequency   order_frequency_flag  
0    Low spender                   20.5  Non-frequent customer  
1    Low spender                   20.5  Non-frequent customer  
2    Low spender                   20.5  Non-frequent customer  
3    Low spender                   20.5  Non-frequent customer  
4    Low spender                   20.5  Non-frequent customer  
..           ...                    ...                    ...  
75   Low spender                   19.0       Regular customer  
76   Low spender                   19.0       Regular customer  
77   Lo

## We have taken sampling data in both orders_products data and in customers data set with 'user_id' as key column and sorted the first 100000 customers with orders for visualizations


In [78]:
filtered_customers = customers_df.loc[(customers_df['user_id'] >=1) & (customers_df['user_id'] <=100000)]

In [86]:
filtered_customers.head(100)

user_id first_name last_name  gender          state  age date_joined  \
0      26711    Deborah  Esquivel  Female       Missouri   48    1/1/2017   
1      33890   Patricia      Hart  Female     New Mexico   36    1/1/2017   
2      65803    Kenneth    Farley    Male          Idaho   35    1/1/2017   
8      69965     Jeremy      Vang    Male          Texas   47    1/1/2017   
9      82820      Shawn     Chung    Male       Virginia   26    1/1/2017   
..       ...        ...       ...     ...            ...  ...         ...   
171    34649      Susan      Nash  Female          Maine   51    1/2/2017   
172    35224        Amy     Gibbs  Female  Massachusetts   67    1/2/2017   
173    36419      Robin   Hendrix  Female     California   70    1/2/2017   
174    43027       Jose    Morgan    Male  Massachusetts   79    1/2/2017   
175    46262     Brenda     Rojas  Female     California   37    1/2/2017   

     num_of_dependants   martital_status  income  
0                    3           married  165665  
1                    0            single   59285  
2                    2           married   99568  
8                    1           married  162432  
9                    2           married   32072  
..                 ...               ...     ...  
171                  1           married   43250  
172                  0  divorced/widowed   94110  
173                  2           married  104030  
174                  3           married  101296  
175                  3           married   52535  

[100 rows x 10 columns]

In [77]:
filtered_orders = ords_prods_merged_df2.loc[ (ords_prods_merged_df2['user_id'] >=1) & (ords_prods_merged_df2['user_id'] <=100000)]

In [80]:
filtered_orders.shape

(15791657, 26)

In [81]:
filtered_customers.shape

(100000, 10)

In [79]:
ords_prods_customer_merge = filtered_orders.merge(filtered_customers, on = ['user_id'])

In [82]:
ords_prods_customer_merge.shape

(15791657, 35)

In [85]:
ords_prods_customer_merge.head(100)

Unnamed: 0_x  order_id  user_id eval_set  order_number  \
0              0   2539329        1    prior             1   
1              1   2398795        1    prior             2   
2              2    473747        1    prior             3   
3              3   2254736        1    prior             4   
4              4    431534        1    prior             5   
..           ...       ...      ...      ...           ...   
95           134   3317979       15    prior             5   
96           136   2685110       15    prior             7   
97           138    887727       15    prior             9   
98           140   2600170       15    prior            11   
99           142   2332087       15    prior            13   

    orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                    2                  8                     NaN         196   
1                    3                  7                    15.0         196   
2                    3                 12                    21.0         196   
3                    4                  7                    29.0         196   
4                    4                 15                    28.0         196   
..                 ...                ...                     ...         ...   
95                   4                 15                    17.0       14715   
96                   1                 11                    17.0       14715   
97                   2                 13                     7.0       14715   
98                   2                  9                    14.0       14715   
99                   1                 10                    14.0       14715   

    add_to_cart_order  reordered _merge  Unnamed: 0_y   product_name  \
0                   1          0   both           195           Soda   
1                   1          1   both           195           Soda   
2                   1          1   both           195           Soda   
3                   1          1   both           195           Soda   
4                   1          1   both           195           Soda   
..                ...        ...    ...           ...            ...   
95                  1          1   both         14715  Coconut Water   
96                  3          1   both         14715  Coconut Water   
97                  1          1   both         14715  Coconut Water   
98                  1          1   both         14715  Coconut Water   
99                  3          1   both         14715  Coconut Water   

    aisle_id  department_id  prices     busiest_day     busiest_days  \
0         77              7     9.0  Regularly busy   Regularly busy   
1         77              7     9.0  Regularly busy  Least busy days   
2         77              7     9.0  Regularly busy  Least busy days   
3         77              7     9.0      Least busy  Least busy days   
4         77              7     9.0      Least busy  Least busy days   
..       ...            ...     ...             ...              ...   
95        98              7     4.0      Least busy  Least busy days   
96        98              7     4.0  Regularly busy     Busiest days   
97        98              7     4.0  Regularly busy   Regularly busy   
98        98              7     4.0  Regularly busy   Regularly busy   
99        98              7     4.0  Regularly busy     Busiest days   

   busiest_period_of_day  max_order      loyalty_flag  average_price  \
0         Average orders         10      New customer       6.367797   
1         Average orders         10      New customer       6.367797   
2            Most orders         10      New customer       6.367797   
3         Average orders         10      New customer       6.367797   
4            Most orders         10      New customer       6.367797   
..                   ...        ...               ...            ...   
95           Most orders         22  Regular custome

#9.Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [88]:
newpath = 'preparedData'

In [89]:
ords_prods_customer_merge.to_pickle(os.path.join(newpath, 'orders_products_customer_merge_spl.pkl'))